File: 2025-02-February-GFG-POTD/February 05 - Mirror Tree/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

struct Node {
    int data;
    Node *left, *right;

    Node(int val) {
        data = val;
        left = right = NULL;
    }
};

class Solution {
public:
    // Function to convert a binary tree into its mirror tree.
    void mirror(Node* node) {
        if (node == NULL) return;

        // Swap the left and right child nodes
        swap(node->left, node->right);

        // Recursively call for left and right subtrees
        mirror(node->left);
        mirror(node->right);
    }
};

// Function to build a tree from level order input
Node* buildTree(vector<string> &ip) {
    if (ip.empty() || ip[0] == "N") return NULL;

    Node* root = new Node(stoi(ip[0]));
    queue<Node*> q;
    q.push(root);

    int i = 1;
    while (!q.empty() && i < ip.size()) {
        Node* curr = q.front();
        q.pop();

        if (ip[i] != "N") {
            curr->left = new Node(stoi(ip[i]));
            q.push(curr->left);
        }
        i++;

        if (i >= ip.size()) break;

        if (ip[i] != "N") {
            curr->right = new Node(stoi(ip[i]));
            q.push(curr->right);
        }
        i++;
    }
    return root;
}

// Function to print level-order traversal of the tree
void levelOrder(Node* root) {
    if (!root) {
        cout << "N\n";
        return;
    }

    queue<Node*> q;
    q.push(root);

    while (!q.empty()) {
        Node* curr = q.front();
        q.pop();

        if (!curr) {
            cout << "N ";
            continue;
        }
        cout << curr->data << " ";
        q.push(curr->left);
        q.push(curr->right);
    }
    cout << endl;
}

int main() {
    int t;
    cin >> t;
    cin.ignore();

    while (t--) {
        string s;
        getline(cin, s);
        vector<string> ip;
        stringstream ss(s);
        string temp;
        while (ss >> temp) {
            ip.push_back(temp);
        }

        Node* root = buildTree(ip);

        Solution obj;
        obj.mirror(root);
        levelOrder(root);
        cout << "~" << endl;
    }

    return 0;
}

File: 2025-02-February-GFG-POTD/February 05 - Mirror Tree/Solution.py
import sys
from collections import deque

class Node:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

class Solution:
    # Function to convert a binary tree into its mirror tree.
    def mirror(self, node):
        if node is None:
            return

        # Swap the left and right child nodes
        node.left, node.right = node.right, node.left

        # Recursively call for left and right subtrees
        self.mirror(node.left)
        self.mirror(node.right)

# Function to build a tree from level order input
def buildTree(s):
    if not s or s[0] == "N":
        return None

    ip = s.split()
    root = Node(int(ip[0]))
    queue = deque([root])

    i = 1
    while queue and i < len(ip):
        curr = queue.popleft()

        if ip[i] != "N":
            curr.left = Node(int(ip[i]))
            queue.append(curr.left)
        i += 1

        if i >= len(ip):
            break

        if ip[i] != "N":
            curr.right = Node(int(ip[i]))
            queue.append(curr.right)
        i += 1

    return root

# Function to print level-order traversal of the tree
def levelOrder(root):
    if root is None:
        print("N")
        return

    queue = deque([root])
    result = []

    while queue:
        curr = queue.popleft()

        if curr is None:
            result.append("N")
            continue

        result.append(str(curr.data))
        queue.append(curr.left)
        queue.append(curr.right)

    print(" ".join(result))

if __name__ == "__main__":
    t = int(sys.stdin.readline().strip())

    for _ in range(t):
        s = sys.stdin.readline().strip()
        root = buildTree(s)

        obj = Solution()
        obj.mirror(root)
        levelOrder(root)
        print("~")
